### PR TITLE
Restore redo functions for AO smgr

### PR DESCRIPTION
**Pull request by usernamedt**
Basically, this PR is analog of the #12785 but opened against the `master` branch since it also needs the change.

### CODE REVIEW

File: src/include/access/xlog.h [hunk 1]
```diff
@@ -379,6 +379,7 @@ extern SessionBackupState get_backup_status(void);
 
 /* Greenplum additions */
 extern bool IsStandbyMode(void);
```
Comment (ashwinstar) [inline src/include/access/xlog.h:381]: Given that only usage of `IsStandbyMode()` is being taken out this PR, please go ahead to remove that function as well from code.

File: src/backend/access/transam/xlog.c [hunk 2]
```diff
@@ -12829,6 +12829,18 @@ IsStandbyMode(void)
 	return StandbyMode;
 }
 
+/*
+ * True if we are running crash recovery.
+ * False if we are running standby continious recovery or not in recovery at all
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:12834]: s/continious/continuous

Comment (ashwinstar) [reply src/backend/access/transam/xlog.c:12834]: I think we should comment as "False if running standby continuous or archive recovery" (no need to mention not in recovery, as then startup process shouldn't even exist)

File: src/backend/access/transam/xlog.c [hunk 3]
```diff
@@ -12829,6 +12829,18 @@ IsStandbyMode(void)
...
+ * This only works in the startup process as the ArchiveRecoveryRequested and
+ * StandbyModeRequested flags are not in shared memory.
+ */
+bool
+IsCrashRecoveryOnly(void)
+{
+	return ArchiveRecoveryRequested == false && StandbyModeRequested == false;
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:12841]: maybe better to write as `!ArchiveRecoveryRequested && !StandbyModeRequested`

**Review by jimmyyih — APPROVED** [review]
+1 on Ashwin's PR notes.  Other than those, the changes look fine.  The simple TAP test is great.

File: src/backend/access/transam/xlog.c (hunk 3 above)
Comment (usernamedt) [reply src/backend/access/transam/xlog.c]: Fixed in https://github.com/greenplum-db/gpdb/pull/12963/commits/aa42192e725855074e4ded6d31ef7ab36d803ada

**Comment by usernamedt** [comment]
I've also applied the improvements to the `IsCrashRecoveryOnly` function in #12785.

**Comment by usernamedt** [comment]
Looks like the CI failed because of the flaky test: https://github.com/greenplum-db/gpdb/pull/12093

**Comment by ashwinstar** [comment]
Pushed this to master branch via 7bc7f6d2544dc25dec4c32b4191d1b9330f76a67. Will push the backport PR after some more due diligence by next week. Thanks.

**Pull request closed without merging**